### DISCUSSION TITLE
Synology NAS (in Docker Container) does not connect to the WireGuard server.

### DISCUSSION THREAD

**Discussion by antegral** [qdm12/gluetun#910] [Q&A] [answered]
I can connect to the WireGuard server from my PC, but I can't connect to the WireGuard server from the Docker Container.

![image](https://user-images.githubusercontent.com/60401462/160278609-2cd78ed6-8149-428c-86ff-9fd542ad6b3c.png)

Docker Inspect
https://pastebin.com/6SQYp0ST

Docker Logs
https://pastebin.com/8r4rT1EF

Session State (WireGuard Server)
![image](https://user-images.githubusercontent.com/60401462/160278507-acf12c64-3088-4f08-8be9-08b89b449516.png)

Server in/outbound Packets
![image](https://user-images.githubusercontent.com/60401462/160278830-6f1b467f-0370-481d-ab54-4eead12a32b2.png)

**Comment by antegral** [comment]
As you can see, it keeps responding to Handshake and sending packets, but the client (container) doesn't recognize it.
Add the following additional information to help you resolve it:

## detailed information

### OS Version
Serverside: `Linux *********** 5.13.0-1018-oracle #22~20.04.1-Ubuntu SMP Wed Feb 16 09:28:40 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
`
Clientside: `Linux *********** 4.4.180+ #41890 SMP Thu Jul 15 03:43:42 CST 2021 x86_64 GNU/Linux synology_apollolake_918+`

### WireGuard Version
```
wg -v
wireguard-tools v1.0.20200513 - https://git.zx2c4.com/wireguard-tools/
```

### `tun.ko` Import Status
```
sudo insmod /lib/modules/tun.ko
insmod: ERROR: could not insert module /lib/modules/tun.ko: File exists
```

### Environment Values (Reference: [here](https://pastebin.com/6SQYp0ST))
```
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
                "VPN_SERVICE_PROVIDER=custom",
                "VPN_TYPE=wireguard",
                "VPN_ENDPOINT_IP=132.***.***.4",
                "VPN_ENDPOINT_PORT=51820",
                "VPN_INTERFACE=tun0",
                "OPENVPN_PROTOCOL=udp",
                "OPENVPN_USER=",
                "OPENVPN_PASSWORD=",
                "OPENVPN_USER_SECRETFILE=/run/secrets/openvpn_user",
                "OPENVPN_PASSWORD_SECRETFILE=/run/secrets/openvpn_password",
                "OPENVPN_VERSION=2.5",
                "OPENVPN_VERBOSITY=1",
                "OPENVPN_FLAGS=",
                "OPENVPN_CIPHERS=",
                "OPENVPN_AUTH=",
                "OPENVPN_PROCESS_USER=",
                "OPENVPN_IPV6=off",
                "OPENVPN_CUSTOM_CONFIG=",
                "WIREGUARD_PRIVATE_KEY=***************************************",
                "WIREGUARD_PRESHARED_KEY=S74P*******************************1YwM=",
                "WIREGUARD_PUBLIC_KEY=S74P*******************************1YwM=",
                "WIREGUARD_ADDRESSES=10.100.100.4/32",
                "SERVER_REGIONS=",
                "SERVER_COUNTRIES=",
                "SERVER_CITIES=",
                "SERVER_HOSTNAMES=",
                "ISP=",
                "OWNED_ONLY=no",
                "PRIVATE_INTERNET_ACCESS_OPENVPN_ENCRYPTION_PRESET=",
                "PRIVATE_INTERNET_ACCESS_VPN_PORT_FORWARDING=off",
                "PRIVATE_INTERNET_ACCESS_VPN_PORT_FORWARDING_STATUS_FILE=/tmp/gluetun/forwarded_port",
                "OPENVPN_CLIENTCRT_SECRETFILE=/run/secrets/openvpn_clientcrt",
                "OPENVPN_CLIENTKEY_SECRETFILE=/run/secrets/openvpn_clientkey",
                "SERVER_NUMBER=",
                "SERVER_NAMES=",
                "FREE_ONLY=",
                "MULTIHOP_ONLY=",
                "FIREWALL=on",
                "FIREWALL_VPN_INPUT_PORTS=",
                "FIREWALL_INPUT_PORTS=",
                "FIREWALL_OUTBOUND_SUBNETS=",
                "FIREWALL_DEBUG=on",
                "LOG_LEVEL=debug",
                "HEALTH_SERVER_ADDRESS=127.0.0.1:9999",
                "HEALTH_ADDRESS_TO_PING=1.1.1.1",
                "HEALTH_VPN_DURATION_INITIAL=30000s",
                "HEALTH_VPN_DURATION_ADDITION=30000s",
                "DOT=on",
                "DOT_PROVIDERS=cloudflare",
                "DOT_PRIVATE_ADDRESS=127.0.0.1/8,10.0.0.0/8,172.16.0.0/12,192.168.0.0/16,169.254.0.0/16,::1/128,fc00::/7,fe80::/10,::ffff:7f00:1/104,::ffff:a00:0/104,::ffff:a9fe:0/112,::ffff:ac10:0/108,::ffff:c0a8:0/112",
                "DOT_VERBOSITY=1",
                "DOT_VERBOSITY_DETAILS=0",
                "DOT_VALIDATION_LOGLEVEL=0",
                "DOT_CACHING=on",
                "DOT_IPV6=off",
                "BLOCK_MALICIOUS=on",
                "BLOCK_SURVEILLANCE=off",
                "BLOCK_ADS=off",
                "UNBLOCK=",
                "DNS_UPDATE_PERIOD=24h",
                "DNS_ADDRESS=1.1.1.1",
                "DNS_KEEP_NAMESERVER=off",
                "HTTPPROXY=",
                "HTTPPROXY_LOG=off",
                "HTTPPROXY_LISTENING_ADDRESS=:8888",
                "HTTPPROXY_USER=",
                "HTTPPROXY_PASSWORD=",
                "HTTPPROXY_USER_SECRETFILE=/run/secrets/httpproxy_user",
                "HTTPPROXY_PASSWORD_SECRETFILE=/run/secrets/httpproxy_password",
                "SHADOWSOCKS=off",
                "SHADOWSOCKS_LOG=off",
                "SHADOWSOCKS_LISTENING_ADDRESS=:8388",
                "SHADOWSOCKS_PASSWORD=",
                "SHADOWSOCKS_PASSWORD_SECRETFILE=/run/secrets/shadowsocks_password",
                "SHADOWSOCKS_CIPHER=chacha20-ietf-poly1305",
                "HTTP_CONTROL_SERVER_ADDRESS=:8000",
                "UPDATER_PERIOD=0",
                "UPDATER_VPN_SERVICE_PROVIDERS=",
                "PUBLICIP_FILE=/tmp/gluetun/ip",
                "PUBLICIP_PERIOD=12h",
                "PPROF_ENABLED=no",
                "PPROF_BLOCK_PROFILE_RATE=0",
                "PPROF_MUTEX_PROFILE_RATE=0",
                "PPROF_HTTP_SERVER_ADDRESS=:6060",
                "VERSION_INFORMATION=on",
                "TZ=Asia/Seoul",
                "PUID=",
                "PGID="
```

**Reply by qdm12** [reply]
Interesting. Yeah it looks like `Received invalid response message from 132.***.***.4:51820` they don't understand each other. I'm away from keyboard right now, but I'll update the Wireguard dependency code tomorrow, maybe that will fix it. I'll message here once it's up.

**Comment by antegral** [comment, marked as answer]
# TL;DR
This problem has been solved with a simple way to **remove the `WIREGUARD_PRESHARED_KEY` environment variable.**

## Why did this happen?
According to [this](https://github.com/qdm12/gluetun/wiki/Custom-provider#tldr), `WIREGUARD_PRESHARED_KEY` exists as an environmental variable.
However, the `WIREGUARD_PRESHARED_KEY` is excluded [here](https://github.com/qdm12/gluetun/wiki/Custom-provider#required-environment-variables).

I specified the environment variable as `WIREGUARD_PRESHARED_KEY` with the same value as `WIREGUARD_PUBLIC_KEY`.
As a result, the handshake with the server did not proceed properly.


## How do you solve this problem?
I excluded `WIREGUARD_PRESHARED_KEY` and completed the authentication information with two environmental variables: `WIREGUARD_PUBLIC_KEY` and `WIREGUARD_PRIVATE_KEY`.

**I think this is a "conflict".**

It seems that the `WIREGUARD_PRESHARED_KEY` is used in wireguard VPN mode.
Anyway, I removed `WIREGUARD_PRESHARED_KEY` from the environment variable and it worked normally.

## if you use Synology NAS?
**use portainer.**

**The Docker GUI functionality supported by Synology NAS is [clearly limited.](https://github.com/qdm12/gluetun/discussions/903#discussioncomment-2405606)**

The most important container network mode (container:gluetun etc...) configuration is not supported,
so it needs to be modified by receiving the json file.

**Reply by qdm12** [reply]
`WIREGUARD_PRESHARED_KEY` is something else than the private or public key. It's a pre-shared key, that is optional, so I think this is all working as expected. I also added `WIREGUARD_PRESHARED_KEY` to the wiki for the custom provider, and changed its value to be different than the private key (yes that was confusing).
